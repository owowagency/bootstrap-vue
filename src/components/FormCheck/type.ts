import {PropType} from 'vue';

const types = ['checkbox', 'radio'] as const;

type Type = typeof types[number];

const typeProp = {
    type: String as PropType<Type>,
    default: 'checkbox',
    validator: (t: Type): boolean => types.includes(t),
};

export {typeProp, Type, types};
