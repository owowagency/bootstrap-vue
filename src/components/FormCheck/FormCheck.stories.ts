import FormCheck from '.';

export default {
    title: 'Forms/Form check',
    component: FormCheck,
    argTypes: {},
};

const template = (args) => ({
    components: {FormCheck},
    setup() {
        return {args};
    },
    template: `
        <FormCheck v-bind="args" />
        <FormCheck v-bind="args" :model-value="true" />
        <FormCheck v-bind="args" disabled />
        <FormCheck v-bind="args" disabled :model-value="true" />
    `,
});

export const Checkbox = template.bind({});

Checkbox.args = {
    label: 'I am a checkbox',
    type: 'checkbox',
};

export const Radio = (args) => ({
    components: {FormCheck},
    setup() {
        return {args};
    },
    template: `
        <FormCheck v-bind="args" />
        <FormCheck v-bind="args" model-value="1" value="1" />
        <FormCheck v-bind="args" disabled />
        <FormCheck v-bind="args" disabled model-value="2" value="2" />
    `,
});

Radio.args = {
    label: 'I am a radio',
    type: 'radio',
};

export const Switch = template.bind({});

Switch.args = {
    label: 'I am a switch',
    switch: true,
};

export const Inline = template.bind({});

Inline.args = {
    inline: true,
    label: 'I am a inline',
};
