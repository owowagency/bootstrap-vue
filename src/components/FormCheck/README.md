# FormCheck

```vue
<FormCheck v-model="check" label="Check me" switch />
```

## Imported Props

| Prop name | Description                   | Type    | Values | Default |
| --------- | ----------------------------- | ------- | ------ | ------- |
| disabled  | Disables the component        | boolean | -      | false   |
| id        | The id used for the component | string  | -      | `uuid`  |

## Props

| Prop name  | Description                                                     | Type                    | Values                | Default   |
| ---------- | --------------------------------------------------------------- | ----------------------- | --------------------- | --------- |
| inline     | Adds the `form-check-inline` class                              | boolean                 | -                     | false     |
| inputClass | The classes that will be forwarded to the input                 | string\|array\|object   | -                     | undefined |
| label      | The label to display                                            | string                  | -                     | undefined |
| modelValue | The model value                                                 | boolean\|string\|number | -                     | false     |
| name       | The name of the input                                           | string                  | -                     | undefined |
| switch     | Display the check as a switch by adding the `form-switch` class | boolean                 | -                     | false     |
| type       | The type of the input                                           | typeProp                | `'checkbox'\|'radio'` |           |
| value      | The value of the input                                          | string\|number          | -                     | undefined |

## Slots

| Name    | Description                       | Bindings |
| ------- | --------------------------------- | -------- |
| default | Displays content inside the label |          |

## Events

| Event name        | Properties                      | Description                           |
| ----------------- | ------------------------------- | ------------------------------------- |
| update:modelValue | **value** `any` - The new value | Fired when the model value is updated |
