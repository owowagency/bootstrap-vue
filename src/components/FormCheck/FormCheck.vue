<template>
    <div
        class="form-check"
        :class="{'form-switch': isSwitch, 'form-check-inline': inline}"
    >
        <input
            :id="id"
            v-model="checked"
            class="form-check-input"
            :class="inputClass"
            :disabled="disabled"
            :name="name"
            :type="type"
            :value="value"
        >

        <label
            v-if="$slots.default || label"
            class="form-check-label"
            :for="id"
        >
            <!-- @slot Displays content inside the label -->
            <slot>
                {{ label }}
            </slot>
        </label>
    </div>
</template>

<script lang="ts">
import {computed} from 'vue';
import {disabledProps} from '@/composables/useDisabled';
import {idProps} from '@/composables/useId';
import {typeProp} from '@/components/FormCheck/type';
</script>

<script lang="ts" setup>
const props = defineProps({
    ...disabledProps,
    ...idProps,
    /**
     * Adds the `form-check-inline` class
     */
    inline: {
        type: Boolean,
        default: false,
    },
    /**
     * The classes that will be forwarded to the input
     */
    inputClass: {
        type: [String, Array, Object],
        default: undefined,
    },
    /**
     * The label to display
     */
    label: {
        type: String,
        default: undefined,
    },
    /**
     * The model value
     */
    modelValue: {
        type: [Boolean, String, Number],
        default: false,
    },
    /**
     * The name of the input
     */
    name: {
        type: String,
        default: undefined,
    },
    /**
     * Display the check as a switch by adding the `form-switch` class
     */
    switch: {
        type: Boolean,
        default: false,
    },
    /**
     * The type of the input
     * @values 'checkbox'|'radio'
     */
    type: typeProp,
    /**
     * The value of the input
     */
    value: {
        type: [String, Number],
        default: undefined,
    },
});

const emit = defineEmits([
    /**
     * Fired when the model value is updated
     * @param {any} value The new value
     */
    'update:modelValue',
]);

const checked = computed({
    get: () => props.modelValue,
    set(v: string | number | boolean | undefined) {
        let value: string | number | boolean | undefined = v;

        if (props.value) {
            value = v ? props.value : undefined;
        }

        emit('update:modelValue', value);
    },
});

// Required to prevent the use of a reserved word.
const isSwitch = computed(() => props.switch);
</script>

<docs>
```vue
<FormCheck
    v-model="check"
    label="Check me"
    switch
/>
```

## Imported Props

| Prop name | Description                   | Type    | Values | Default |
| --------- | ----------------------------- | ------- | ------ | ------- |
| disabled  | Disables the component        | boolean | -      | false   |
| id        | The id used for the component | string  | -      | `uuid`  |
</docs>
