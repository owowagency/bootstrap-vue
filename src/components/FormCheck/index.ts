import FormCheck from './FormCheck.vue';
export default FormCheck;
export * from './type';
