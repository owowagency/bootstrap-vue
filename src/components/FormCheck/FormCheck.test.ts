import {describe, expect, it} from 'vitest';
import FormCheck from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(FormCheck, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentRenderTest(
        FormCheck,
        // Force an id to prevent snapshots from failing because of uuid.
        {
            props: {
                id: 'dont-change-please',
                label: 'I am a label',
            },
        },
        true,
        'renders label',
    );

    componentSlotRenderTest(FormCheck);

    it('renders radio input type', () => {
        const wrapper = shallowMount(FormCheck, {props: {type: 'radio'}});

        const input: HTMLInputElement = wrapper.find('input').element;

        expect(input.type).toBe('radio');
    });

    componentWrapperClassTest(FormCheck, {switch: true}, 'form-switch');

    componentWrapperClassTest(FormCheck, {inline: true}, 'form-check-inline');

    it('adds input class to input', () => {
        const wrapper = shallowMount(FormCheck, {props: {inputClass: 'is-invalid'}});

        const input = wrapper.find('input');

        expect(input.classes()).toContain('is-invalid');
    });

    it('emits on checked change', async() => {
        const wrapper = shallowMount(FormCheck);

        wrapper.find('input').setValue(true);

        expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    });
});

describe('checked', () => {
    it('is false by default', () => {
        const wrapper = shallowMount(FormCheck);

        expect(wrapper.vm.checked).toBe(false);
    });

    it('is true when given', () => {
        const wrapper = shallowMount(FormCheck, {
            props: {modelValue: true},
        });

        expect(wrapper.vm.checked).toBe(true);
    });

    it('is equal to modelValue', () => {
        const wrapper = shallowMount(FormCheck, {
            props: {modelValue: 'b'},
        });

        expect(wrapper.vm.checked).toBe('b');
    });

    it('emit boolean when no value', () => {
        const wrapper = shallowMount(FormCheck);

        wrapper.vm.checked = true;

        expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual([true]);

        wrapper.vm.checked = false;

        expect(wrapper.emitted('update:modelValue')[1]).toStrictEqual([false]);
    });

    it('emits value when given', () => {
        const wrapper = shallowMount(FormCheck, {
            props: {value: 'a'},
        });

        wrapper.vm.checked = true;

        expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual(['a']);

        wrapper.vm.checked = false;

        expect(wrapper.emitted('update:modelValue')[1]).toStrictEqual([undefined]);
    });
});
