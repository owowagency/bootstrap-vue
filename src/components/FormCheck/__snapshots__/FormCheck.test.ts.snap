// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"form-check\\"><input id=\\"dont-change-please\\" class=\\"form-check-input\\" type=\\"checkbox\\">
  <!--v-if-->
</div>"
`;

exports[`template > renders label 1`] = `
"<div class=\\"form-check\\"><input id=\\"dont-change-please\\" class=\\"form-check-input\\" type=\\"checkbox\\"><label class=\\"form-check-label\\" for=\\"dont-change-please\\">
    <!-- @slot Displays content inside the label -->I am a label
  </label></div>"
`;
